### ISSUE TITLE
[JENKINS-57528] Jenkins in Docker does not install detached plugins when there is no UC data

### ISSUE THREAD

**Issue by timja-bot** [timja-org/jenkins-gh-issues-poc-11-04#4778]
<p>It is a placeholder for <a href="https://github.com/jenkinsci/docker/issues/698" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/docker/issues/698</a> and for <a href="https://github.com/jenkinsci/jenkins/pull/4000" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/pull/4000</a> which addresses it in the core</p>

<p>Using newer cores that have part of it moved to plugins and is now implied dependencies in other plugins is causing to have bad Jenkins installation.</p>

<p> </p>

---
<details><summary><i>Originally reported by <a title="Oleg Nenashev" href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=oleg_nenashev">oleg_nenashev</a>, imported from: <a href="https://issues.jenkins.io/browse/JENKINS-57528" target="_blank">Jenkins in Docker does not install detached plugins when there is no UC data</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Jesse Glick" href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick">jglick</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Critical
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2019-05-17T14:28:44+00:00
<li><b>imported</b>: 2025-11-05
</ul></i>
</details>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nrayapati">nrayapati</a>:</i>
<p>And also other test use case failing with similar errors. For some instances, we test Jenkins locally before we apply any changes to actual instance, part of that we have script which copies over all the hpi files to plugins directory before we start up the jenkins, this is to make sure these new plugins don't have any security issues and also the dependencies look good.  </p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=oleg_nenashev">oleg_nenashev</a>:</i>
<p>It is not this issue per se, but it is "Bring the bundled version of the Script Security plugin up to date with recent security advisories, in the unlikely case it is indeed installed from the WAR rather than the update center" which was delivered along with it in <a href="https://github.com/jenkinsci/jenkins/pull/4000" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/pull/4000</a> . It looks like the code triggered the downgrade somehow.</p>

<p>CC <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick" class="user-hover" rel="jglick">jglick</a> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=danielbeck" class="user-hover" rel="danielbeck">danielbeck</a> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=batmat" class="user-hover" rel="batmat">batmat</a></p>

<p>A separate issue would be appreciated</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick">jglick</a>:</i>
<blockquote><p>script which copies over all the hpi files to plugins directory</p></blockquote>

<p>Note that <tt>$JENKINS_HOME/plugins/</tt> is expected to contain <tt>&#42;.jpi</tt> files, never <tt>&#42;.hpi</tt> which is the extension used in Maven repositories and update center URLs. The Jenkins plugin manager always performs this rename when installing plugins either from an update center or the <tt>install-plugin</tt> CLI command, and official packaging scripts such as for Docker do the same, but perhaps your unsupported custom script neglected to do that? If this is the issue, then Jenkins core could be made a bit more robust here and either silently tolerate <tt>&#42;.hpi</tt> in more cases, print a clear warning about improper file names, or try to automatically clean up. That would be a separate issue that could be linked to this one.</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick">jglick</a>:</i>
<p>The files in the <tt>plugins</tt> directory are expected to be named <tt>ARTIFACT_ID.jpi</tt>, so for example <tt>script-security.jpi</tt> rather than <tt>script-security-1.58.hpi</tt> (filename in Artifactory) or <tt>script-security.hpi</tt> (filename in updates.jenkins.io).</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick">jglick</a>:</i>
<p>I filed a refinement in jenkins #4039 to improve the behavior and make it more apparent what you were doing wrong.</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nrayapati">nrayapati</a>:</i>
<p>Thank you very much <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick" class="user-hover" rel="jglick">jglick</a> we changed our script to use .jpi instead of .hpi and also removed the version from the file name. It worked! </p>

<p>Thanks again really appreciate your inputs. </p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick">jglick</a>:</i>
<p>There was an ATH regression which it is hoped form-element-path #8 will help correct.</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=raspy">raspy</a>:</i>
<p>Is it possible to somehow skip or remove this behavior? We do not use any of those detached plugins and they keep being instantiated (in rather ancient versions) at service startup.</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick">jglick</a>:</i>
<blockquote><p>We do not use any of those detached plugins</p></blockquote>

<p>If you are using the official Docker image as a base, you ought to be able to</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-none">RUN touch /usr/share/jenkins/ref/plugins/{whatever-id,another-id}.jpi.disabled
</pre>
</div></div>

<p>Untested, YMMV.</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=awiddersheim">awiddersheim</a>:</i>
<p>This feels a little bit untenable if I'm understanding everything correctly, at least for those using the Docker images. Like I'll be playing whack-a-mole either by adding these implied dependencies to my list of installed plugins so they get updated (because of security issues) or by disabling them because they aren't actually used but are pulling in ancient versions with tons of security warnings. It also heavily relies on plugin maintainers to be constantly updating the version of Jenkins their plugin relies on to not be continually pulling in implied deps of old Jenkins core versions.</p>

<p>For example, the simple <a href="https://plugins.jenkins.io/purge-build-queue-plugin#" class="external-link" target="_blank" rel="nofollow noopener">https://plugins.jenkins.io/purge-build-queue-plugin#</a> is referencing a really old Jenkins core version so it's pulling in a ton of implied deps, most of which are really old and some with security vulnerabilities. Whats the best way to handle this other than having that plugin updated to reference a new Jenkins version so less implied deps are pulled in?</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=awiddersheim">awiddersheim</a>:</i>
<p>Er, I'm sorry. The <a href="https://plugins.jenkins.io/pam-auth" class="external-link" target="_blank" rel="nofollow noopener">https://plugins.jenkins.io/pam-auth</a> plugin is the one with the security issue.</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">jenkins_1  | INFO: Loading a detached plugin as a dependency: /<span class="code-keyword">var</span>/jenkins_home/plugins/pam-auth.jpi
jenkins_1  | WARNING: Created /<span class="code-keyword">var</span>/jenkins_home/plugins/pam-auth/WEB-INF/lib/classes.jar; update plugin to a version created with a newer harness
jenkins_1  | INFO: Took 0ms <span class="code-keyword">for</span> Loading plugin PAM Authentication plugin v1.1 (pam-auth) by pool-6-thread-4
jenkins_1  | INFO: Took 0ms <span class="code-keyword">for</span> Initializing plugin pam-auth by pool-6-thread-1 </pre>
</div></div>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick">jglick</a>:</i>
<p>Then Jenkins should be bundling 1.5.1.</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=danielbeck">danielbeck</a>:</i>
<p>I'll take this.</p>

<p> </p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=awiddersheim">awiddersheim</a>:</i>
<p>Thanks <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=danielbeck" class="user-hover" rel="danielbeck">danielbeck</a>. Did you need me to make a ticket or did you already?</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=awiddersheim">awiddersheim</a>:</i>
<p>Created <a href="https://issues.jenkins-ci.org/browse/JENKINS-59552" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-59552</a>.</p>
